### ISSUE TITLE
newline in zsh after 0.7.0

### ISSUE THREAD

**Issue by d4be4st** [piotrmurach/tty-prompt#36]
in versoin 0.7 onwards after using anything in tty-prompt my zsh shell starts adding  newlines after return:

0.6.0 version:
![0.6.0 version](https://scr.infinum.co/stjepan_hadjic/1._stefStefs-MacBook-Pro_devgemstty_test_zsh_2017-01-19_09-58-01.png)

0.7.0 version:
![0.7.0 version](https://scr.infinum.co/stjepan_hadjic/1._stefStefs-MacBook-Pro_devgemstty_test_zsh_2017-01-19_09-58-36.png)

I believe these changes to reader mode have somehow introduced this weird bug: https://github.com/piotrmurach/tty-prompt/compare/v0.6.0...v0.7.0#diff-d0c27415e6027553e90af174f4d24c98

I do not have enough shell knowledge to figure out what

I tried the same example in bash and sh and the bug is not present there.

Here is a screenshoot of my collegues zsh setup:
![0.7.0 version](https://scr.infinum.co/adis/adisJobs__zsh_2017-01-19_10-05-52.jpg)

**Comment by piotrmurach** [comment]
I don't use zsh myself and this seems like a super weird bug, world of shells seems to be super inconsistent. I'm currently working on Windows support as a priority. I would appreciate help figuring out this problem. The only thing that significantly changed between `0.6.0` and `0.7.0` was how the terminal modes are handled(echo, raw etc...) by switching over to use `io/console` standard library. Could you please try testing off the master branch and point your `tty-prompt` dependency at this git repo?

**Comment by carlosefonseca** [comment]
Hi. This happens to me too, with zsh…
I installed the master version but still happens.
<img width="242" alt="screen shot 2017-02-15 at 20 42 55" src="https://cloud.githubusercontent.com/assets/14120/22994591/630cbfa8-f3bf-11e6-9052-05f9a1ae4f0c.png">
Any idea of what else could we test?

**Comment by piotrmurach** [comment]
Thanks for confirming the bug. Unfortunately I don't have an image ready with zsh to test. I'd really like to fix this but it will need to wait a bit unless of course I get a bit of help solving this mystery. 

The [changelog](https://github.com/piotrmurach/tty-prompt/blob/master/CHANGELOG.md) documents all the significant changes. If you select given version you will see the full diff like this [one](https://github.com/piotrmurach/tty-prompt/compare/v0.6.0...v0.7.0). I think looking through the commits you may be able to spot the problem. The most significant change is how console modes are handled. My gut feeling is that, something must have changed with how `echo` and `raw` modes are dealt with. 

I think it would be worth checking also [examples](https://github.com/piotrmurach/tty-prompt/tree/master/examples). It may well be that this affects a certain type of prompt, which would narrow the problem as different prompts read characters differently.

**Comment by d4be4st** [comment]
I think i found it.

https://github.com/piotrmurach/tty-prompt/blob/master/lib/tty/prompt/reader/mode.rb#L22
For some reason this is true for my $stdin. And because of this line https://github.com/piotrmurach/tty-prompt/blob/master/lib/tty/prompt/reader/mode.rb#L26 my stdin thinks it needs to echo everything afterwards.

I have put false there and it stopped repeating the lines.

As for the solution, I have no idea, seems like a bug in ruby itself...

**Comment by piotrmurach** [comment]
Thanks for finding this bug. Let's stump it out :smile:

I would suggest the following:
```ruby
def echo(is_on = true, &block)
  if is_on
    yield
  else
    @input.noecho(&block)
  end
end
```
Would you mind testing this out if it fixes the issue. In case it does, I would really like you to submit PR?

**Comment by d4be4st** [comment]
Your suggestion works on my machine :)

**Comment by piotrmurach** [comment]
Magic! Good for me. Thanks for investigating this. 

I will release new version shortly, probably this weekend.

**Comment by piotrmurach** [comment]
Released `v0.11.0`. Enjoy!
